### PR TITLE
Add repeat function

### PR DESCRIPTION
**Pull request by kmaehashi**
Implemented `F.repeat` (Variable version of `xp.repeat`).

This feature was requested in #3632.
It is also needed to migrate `prod` function to FunctionNode.

### CODE REVIEW

File: chainer/functions/__init__.py [hunk 1]
```diff
@@ -64,6 +64,8 @@
 from chainer.functions.array.pad_sequence import PadSequence  # NOQA
 from chainer.functions.array.permutate import permutate  # NOQA
 from chainer.functions.array.permutate import Permutate  # NOQA
+from chainer.functions.array.repeat import repeat  # NOQA
+from chainer.functions.array.repeat import Repeat  # NOQA
```
Comment (niboshi) [inline chainer/functions/__init__.py]: Please remove `Repeat`. We're not going to expose new function classes.

File: chainer/functions/array/repeat.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            raise ValueError('all elements in repeats must be zero or larger')
+
+        self.axis = axis
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+
+    def forward(self, inputs):
+        self.retain_inputs((0,))
+        x = inputs[0]
```
Comment (niboshi) [inline chainer/functions/array/repeat.py]: This is better: `x, = inputs`

File: chainer/functions/array/repeat.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        xp = cuda.get_array_module(x)
+        repeats = self.repeats
+        if self.axis is None or len(self.repeats) == 1:
+            repeats = self.repeats[0]
+        return xp.repeat(x, repeats, self.axis),
+
+    def backward(self, indexes, grad_outputs):
+        x = self.get_retained_inputs()[0]
```
Comment (niboshi) [inline chainer/functions/array/repeat.py]: Dittoo. `x, = self.get_retained_inputs()`

File: chainer/functions/array/repeat.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            The number of times which each element of ``x`` is repeated.
+        axis (:class:`int`):
+            The axis along which to repeat values.
+
+    Returns:
+        ~chainer.Variable: The repeated output Variable.
+
+    .. admonition:: Example
+
+        >>> x = np.array([0, 1, 2])
```
Comment (niboshi) [inline chainer/functions/array/repeat.py]: This looks NumPy example.

File: chainer/functions/array/repeat.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+class RepeatGrad(function_node.FunctionNode):
+
+    def __init__(self, repeats, axis, in_shape, in_dtype):
+        self.repeats = repeats
+        self.axis = axis
+        self.in_shape = in_shape
+        self.in_dtype = in_dtype
+
+    def forward(self, inputs):
+        gy = inputs[0]
```
Comment (niboshi) [inline chainer/functions/array/repeat.py]: Ditto. `gy, = inputs`

**Comment by kmaehashi** [comment]
Thanks for review, fixed accordingly.

File: chainer/functions/array/repeat.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+from chainer import function_node
+from chainer.utils import type_check
+
+
+class Repeat(function_node.FunctionNode):
+
+    """Repeat elements of an array."""
+
+    def __init__(self, repeats, axis=None):
+        if isinstance(repeats, six.integer_types):
```
Comment (niboshi) [inline chainer/functions/array/repeat.py]: This condition matches if `repeats` is `bool`. Please exclude such cases.
Also, please test such cases in the unit test.

File: chainer/functions/array/repeat.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            self.repeats = (repeats,)
+        elif isinstance(repeats, tuple) and all(
+                isinstance(x, six.integer_types) for x in repeats):
```
Comment (niboshi) [inline chainer/functions/array/repeat.py]: ditto.

File: chainer/functions/array/repeat.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            self.repeats = repeats
+        else:
+            raise TypeError('repeats must be int or tuple of ints')
+
+        if not all(x >= 0 for x in self.repeats):
+            raise ValueError('all elements in repeats must be zero or larger')
+
+        self.axis = axis
```
Comment (niboshi) [inline chainer/functions/array/repeat.py]: How about checking the type and value of `axis`, and adding to the unit test as well?

File: chainer/functions/array/repeat.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 1)
+
+    def forward(self, inputs):
+        self.retain_inputs((0,))
+        x, = inputs
+        xp = cuda.get_array_module(x)
+        repeats = self.repeats
+        if self.axis is None or len(self.repeats) == 1:
+            repeats = self.repeats[0]
```
Comment (niboshi) [inline chainer/functions/array/repeat.py]: Only the first element of `repeats` is used if `axis` is `None`.
That seems to be different from `numpy.repeat`'s behavior.

Comment (kmaehashi) [reply chainer/functions/array/repeat.py]: Sorry, I should have put blocked-by-another-pr label.
During the implementation of this function, I found an issue that cupy & numpy repeat function behaves differently. cupy/cupy#670
This code cannot be fixed until the above PR is merged.

**Comment by kmaehashi** [comment]
This PR needs to be updated after https://github.com/cupy/cupy/pull/670 got merged.

File: chainer/functions/array/repeat.py (hunk 9 above)
Comment (niboshi) [reply chainer/functions/array/repeat.py]: OK.

File: chainer/functions/array/repeat.py (hunk 6 above)
Comment (kmaehashi) [reply chainer/functions/array/repeat.py]: Actually NumPy/CuPy accepts `bool` for `repeats`. Should it be more strict than backend libraries?

```
>>> np.repeat(np.array([1.0, 2.0, 3.0]), True)
array([ 1.,  2.,  3.])
>>> np.repeat(np.array([1.0, 2.0, 3.0]), [False, True, 2])
array([ 2.,  3.,  3.])
```

Comment (niboshi) [reply chainer/functions/array/repeat.py]: Oh... thank you for confirmation. So how about testing those inputs as well?

Comment (kmaehashi) [reply chainer/functions/array/repeat.py]: I agree, I added tests for bools.

File: chainer/functions/array/repeat.py (hunk 8 above)
Comment (kmaehashi) [reply chainer/functions/array/repeat.py]: Exactly, fixed as suggested.

File: chainer/functions/array/repeat.py (hunk 9 above)
Comment (kmaehashi) [reply chainer/functions/array/repeat.py]: I fixed the code assuming https://github.com/cupy/cupy/pull/670 got merged.

**Comment by kmaehashi** [comment]
Jenkins test this please

**Comment by niboshi** [comment]
Is this PR still WIP?
If not, let's remove [WIP] first.

**Comment by kmaehashi** [comment]
Sorry to bother you. 
I have to update test cases to workaround NumPy 1.9 issue (found during fixing cupy.repeat bug): https://github.com/cupy/cupy/pull/670#issuecomment-351902672

I can remove WIP after that.

**Comment by niboshi** [comment]
Ah, OK!

**Comment by kmaehashi** [comment]
@niboshi san, this is ready for review.

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by niboshi** [comment]
LGTM!
I'm sorry for taking time to review.

**Pull request merged**